### ISSUE TITLE
(infinite) recursion in "lua_xmove constructor" of basic_table_core

### ISSUE THREAD

**Issue by rohlem** [ThePhD/sol2#509]
(update: seems I missed the last 3 commits since commit 72143a494b25c2df614c3952a81563d5eebdb5b7 , but they don't affect this part of the code /issue from what I could tell)

I just updated Visual Studio (to Community 2017 15.3.4) and Sol (to 2.18.3) and got the following error when trying the "lua_xmove constructor" of sol::table (as documented [here](http://sol2.readthedocs.io/en/latest/threading.html))

`1>...\sol\sol.hpp(17102): warning C4717: 'sol::basic_table_core<0,sol::reference>::basic_table_core<0,sol::reference><sol::basic_table_core<0,sol::reference> const &,0>': recursive on all control paths, function will cause runtime stack overflow`

(contrived minimal code example to reproduce)
```
sol::state s;
sol::table t(s, sol::table{});
```

While it wouldn't be the first time for an incorrect warning, after looking at the code I for once have to agree with VS on this:

```
	template <typename T, meta::enable_any</*reference or stack_reference, but irrelevant*/> = meta::enabler>
	basic_table_core(lua_State* L, T&& r)
	: basic_table_core(L, std::forward<T>(r)) {}
```

Looking at the implementation in `basic_function` that's used in the documentation (and codebase) example it looks like it's meant to forward to its base type, which I think should look something like `basic_object_base<base_type>(L, std::forward<T>(r))` (with which `basic_object_base` in turn forwards it directly to `base_type`).

(Btw, these "lua_xmove constructors" are currently not documented on the API pages of the respective types, or even on [sol::object's page](http://sol2.readthedocs.io/en/latest/api/object.html) as a start; users have to explicitly find the page on threading. UPDATE: just noticed the reference on [reference's page](http://sol2.readthedocs.io/en/latest/api/reference.html), that's nice at least; hopefully people find that)

**Comment by rohlem** [comment]
Actually, there might be more to it, now I'm experiencing weird behaviour of the "copy's" registry index being LUA_NOREF. I'm currently trying to get a minimal reproducible case of that to then look into which types are effected (or whether it's something completely unrelated).

**Comment by rohlem** [comment]
OK, the second issue was completely separate and has been solved.
( It was that `sol::thread t(original_state);` does not create a new thread, it tries to pop original_state's stack. So you most likely get a nullptr-thread cannot be xmoved to. )
So the only real issue is the incorrect constructor that is described in the first post.

**Comment by ThePhD** [comment]
Fixed and backported into v2.18.3.
